# Sequencing the draf genome of the microscopic _N. westbladi_ from a single worm
## Project summary:
The steady development (and standardization) of sequencing technologies has open new avenues in evolutionary biology, leading to a better understanding about how the diversity of life has been generated and maintained over time. However, producing high-quality genomes remains a challenging task for many animal groups, such as miniaturised animals due to the low DNA yields generated from a typical extraction. This is the case, for instance, of xenacoelomorph worms. Xenacoelomorpha is a 500 million years old phylum of morphologically simple worms, which have become an important model to study the evolution of [body regeneration](https://www.science.org/doi/full/10.1126/science.aau6173), [photosymbiosis](photosymbiosis), or [the nervous system](https://www.nature.com/articles/nature25030), among others. However, due to their millimetric size, only species of large size or that can be cultured in the lab (thus pooling individuals)  have been succesfully sequenced (see: [1](https://www.science.org/doi/full/10.1126/science.aau6173), [2](https://academic.oup.com/gigascience/article/8/4/giz023/5429687?login=true), [3](https://academic.oup.com/g3journal/article/13/2/jkac336/6948452?login=true), [4](https://www.biorxiv.org/content/10.1101/2022.06.24.497508v2.full)).

Here, we have combined the PacBio Ultra-Low Input DNA protocol and HiFi sequencing to produce the first draft genome of Nemertodermatide. We produce this genome from a single, microscopic _Nemertoderma westbladi_ worm, caught in the wild. We show the potential of this approach to generate relatively high-quality genomes from microscopic animals. We observe that genome contiguity remains challenging, but genome quality is good enough to annotate full genes. We demonstrate this by analysing the evolution of the molecular machinery of the excretory system.

## Repository description:
This repository contains a step by step description of the analyses, from read filtering and assembly to comparative genomics. All the relevant data have been released in different repositories (linked below), but the intermediate files necessary to carry out the analyses are also provided.

### Data availability:
<ul>
    <li><strong>Raw sequencing data:</strong> <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=ON951339:ON951583[accn]">Illumina</a> and <a href="https://www.ncbi.nlm.nih.gov/nuccore/AY588229">HiFi</a> reads.</li>
    <li><strong>Genome assembly</strong> is available through the <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=ON968966%3AON968984%5Baccn%5D">NCBI database</a>.</li>
    <li><strong>All intermediate files</strong>, including BUSCO reports and annotations, have been uploaded to <a href="https://www.ncbi.nlm.nih.gov/sra/?term=PRJNA856832">Dryad</a>.</li>
</ul>

## Citation
<ul>
  <li>Abalde, S., Tellgren-Roth, C., Heintz, J., Vinnere-Pettersson, O., and Jondelius, U. (submitted). TITLE. <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=ON951339:ON951583[accn]">bioRxiv</a>.</li>
</ul>

---
